src/ProjectSystem/Impl/Extensions/ProjectTreeExtensions.cs

```diff
@@ -63,6 +63,38 @@ public static class ProjectTreeExtensions {
             return Enumerable.Empty<string>();
         }
 
+        public static IEnumerable<string> GetAllFilePaths(this IEnumerable<IProjectTree> nodes) {
+            if (nodes != null) {
+                List<string> paths = new List<string>();
+                foreach (IProjectTree node in nodes) {
+                    if (node.IsFolder) {
+                        paths.AddRange(node.GetAllFilePaths());
+                    } else if (node.Children.Count > 0) {
+                        paths.AddRange(node.Children.GetAllFilePaths());
+                    } else if (!string.IsNullOrWhiteSpace(node.FilePath)) {
+                        paths.Add(node.FilePath);
+                    }
+                }
+                return paths;
+            }
+            return Enumerable.Empty<string>();
+        }
+
+        public static IEnumerable<string> GetAllFilePaths(this IProjectTree node) {
```

Comment:
Also, we have couple of extension methods that can be useful: `EnumerableExtensions.TraverseBreadthFirst` and `EnumerableExtensions.TraverseDepthFirst`

---
In reply to: [80771759](https://github.com/Microsoft/RTVS/pull/2405#discussion_r80771759) [](ancestors = 80771759,80537483,80395305)